### ISSUE TITLE
Support dynamic quantization config to improve mixtral perplexity?

### ISSUE THREAD

**Issue by lavawolfiee** [mobiusml/hqq#2]
Hi! First of all, thank you for publishing the HQQ algorithm insightful blog post. It was a pleasure to tinker with an algorithm that quantizes the model in minutes instead of hours.
We’d appreciate it if you could comment on some of our observations about HQQ performance.

We tried several quantization regimes for [Mixtral-8x7b-Instruct-v0.1](https://huggingface.co/mistralai/Mixtral-8x7B-Instruct-v0.1) models and your pre-quantized checkpoints from [huggingface hub](https://huggingface.co/mobiuslabsgmbh). To recall, the mixtral model consists of ~45B mlp “experts” and ~1.5B other parameters (attention, gates, embeddings, logits, etc).
In your checkpoints, it appears that you are quantizing all linear layers (except logits and MoE gates) with the same quantization config, e.g. 2-bit or 4-bit.

From our experiments, it appears that the best quality-to-size trade-offs come from **quantizing experts coarsely, but keeping non-expert parameters in higher precision**. We tried several combinations of both HQQ and NF4 algorithms separately for experts and non-expert (attention layers) to the following results.

The table below reports the perplexity on a slice of wikitext calibration data (not the same as wiki2 test set), but it is large enough to be statistically significant.
The “Experts” column has the quantization regime for the [MixtralBLockSparseTop2MLP](https://github.com/huggingface/transformers/blob/388fd314d8dd6bcb685f5223a7a46caa17e577f2/src/transformers/models/mixtral/modeling_mixtral.py#L664) layers.

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th></th>
      <th>Pexplexity</th>
      <th>Memory, GB</th>
    </tr>
    <tr>
      <th>Stem</th>
      <th>Experts</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th rowspan="5" valign="top">fp16</th>
      <th>nf4</th>
      <td>4.24</td>
      <td>25.76</td>
    </tr>
    <tr>
      <th>hqq_4bit_g128</th>
      <td>4.25</td>
      <td>26.82</td>
    </tr>
    <tr>
      <th>hqq_3bit_g128</th>
      <td>4.51</td>
      <td>22.59</td>
    </tr>
    <tr>
      <th>hqq_2bit_g16_s</th>
      <td>4.89</td>
      <td>20.21</td>
    </tr>
    <tr>
      <th>hqq_2bit_g128</th>
      <td>6.87</td>
      <td>15.54</td>
    </tr>
    <tr>
      <th rowspan="5" valign="top">nf4</th>
      <th>nf4</th>
      <td>4.29</td>
      <td>23.75</td>
    </tr>
    <tr>
      <th>hqq_4bit_g128</th>
      <td>4.30</td>
      <td>24.80</td>
    </tr>
    <tr>
      <th>hqq_3bit_g128</th>
      <td>4.57</td>
      <td>20.58</td>
    </tr>
    <tr>
      <th>hqq_2bit_g16_s</th>
      <td>4.98</td>
      <td>18.20</td>
    </tr>
    <tr>
      <th>hqq_2bit_g128</th>
      <td>7.14</td>
      <td>13.53</td>
    </tr>
  </tbody>
</table>

Curiously, if we consider the published [2bit g16 s128 configuration](https://huggingface.co/mobiuslabsgmbh/Mixtral-8x7B-v0.1-hf-2bit_g16_s128-HQQ), the original checkpoint scores ~5.85 on our validation split at 18.03GB model size. However, **if we switch attention layers from 2-bit HQQ to NF4 (~4-bit), we observe 4.98 perplexity with about 1.5% increase in model size (18.2GB)**.

Does this agree or disagree with your intuition about how HQQ should work for mixtral?

If it does, would you consider switching to this hybrid coarse/fine quantization scheme for Mixtral by default or as a built-in functionality of the hqq library?

If possible, we’d also appreciate it if you can verify these results on your side, if you have time.

The code for reproducing the values in the table can be found at this  link: https://gist.github.com/dvmazur/c4d32c753e06d8fc9a78606f84224176
We also include the versions of the main libraries we used in the readme of that gist.

Yours,
Artem Eliseev (myself @lavawolfiee) and Denis Mazur ( @dvmazur )

**Comment by mobicham** [comment]
First for all, thank you very much for taking the time to do this, it's really great! 

Regarding support for dynamic quantization settings, the code already supports this via the patching function by providing custom settings for each layer tag. But in the first release we just wanted to keep things simple for the user so they only have to provide the settings in a single line. 
With these new findings, we will add support for this shortly so the user can feed a dictionary in the form `{layer_tag: BaseQuantizeConfig()} `

Now regarding the benchmarks, here are a couple of comments: 

### HQQ vs. NF4

#### GPU Memory: 

I see that you count the number of bits to determine the memory. A more accurate way would be to clean-up (`torch.cuda.empty_cache() + gc.collect()`) after prediction, then use nvidia-smi. 
This would provide a more accurate GPU VRAM usage for the whole environment because some things are allocated only on prediction time.

#### Settings:

The numbers you provided compare HQQ with a group-size of 128 and partially compressed meta-data vs. NF4 with a group-size of 64 and fully compressed meta-data (default settings). 

For a fair comparison, we should use the same group-size and comparable meta-data compression settings. 

It turns out, the extra memory is because of the meta-data compression, here are the numbers (ppl / VRAM)

No meta-data compression:          
| Method   |      PPL      |  Memory  |
|----------|:-------------:|------:| 
FP4_g64 |               3.9594 | 27313 MB
NF4_g64 |               3.8144 | 27313 MB
HQQ_4-bit_g64 |         3.7914 | 27173 MB 

With meta-data compression: 
| Method   |      PPL      |  Memory  |
|----------|:-------------:|------:| 
FP4_g64_cs |            3.9586 | 25133 MB 
NF4_g64_cs|            3.8159 | 25133 MB   
HQQ_4-bit_g64_z|       3.7936 | 26513 MB 
HQQ_4-bit_g64_z_s256|  3.7942 | 25929 MB 

_HQQ_4-bit_g64_z: zero-point quantized to 8-bit (default setttings)._
_HQQ_4-bit_g64_z_s256: zero-point quantized to 8-bit + scale quantized to 8-bit with g=256)._

In terms of perplexity, HQQ outperforms both FP4 and NF4 for the same group-size, with and without meta-data compression. 
Without meta-data compression, HQQ also uses slightly less memory. With compression, NF4 uses ~800 MB less memory.
I haven't experimented too much but with a more aggressive compression of the zero/scale, but that seems to be the source of the extra memory usage.

Strangely enough, even using no grouping for compressing the scaling parameters, the memory usage stays at 25929 MB, so maybe there's some memory issue somewhere.

###  Dynamic settings: HQQ + NF4 mix
Now let's evaluate how the HQQ/NF4 mixing works for the Mixtral model:

| Method   |      PPL      |  Memory  | Comment
|----------|:-------------:|------:| ------:| 
NF4_g64_cs (att) / HQQ_4-bit_g64_z_s256 (experts)| 3.8011 | 25929 MB | worse than HQQ_4-bit_g64_z_s256
HQQ_4-bit_g64_z_s256 (att) / NF4_g64_cs (experts)| 3.8092 | 25161 MB  | ~better than NF4_g64_cs

For 4-bit HQQ + NF4, it looks like using HQQ for the attention is better than using NF4, As HQQ is more accurate than NF4, this is kinda consistent with the idea that attention layers should use more accurate quantization than the experts. Though, the difference is not extremely large.

###  Dynamic settings: mixed settings for attention/experts:
Now the really interesting part: keeping the attention module in 4-bit while using 2-bit for the experts is indeed a much better setting: same memory usage but a huge drop in perplexity!

| Method   |      PPL      |  Memory  |
|----------|:-------------:|------:| 
HQQ_2bit_g16_z_s128 (att + experts) |                        5.9027 | ~20 GB   
NF4_g64_cs (att)   / HQQ_2bit_g16_z_s128 (experts) |         4.749  | ~20 GB
HQQ_4-bit_g64_z_s256 (att)/ HQQ_2bit_g16_z_s128 (experts) |  4.6864 | ~20 GB             

HQQ vs. NF4 for the attention module is also better for about the same memory usage.

The mixed HQQ 2/4bit model on disk is 17GB while on memory it takes 20GB, something somewhere is leaking, will need to investigate.

Here's the code I used to reproduce all these numbers: https://gist.github.com/mobicham/1f002b54d37ab1809adc5a498670e7f3#file-hqq_nf4_mixtral-py

Thanks a lot guys for pointing this out. We will put this mixed 2/4 version of Mixtral models on Hugging Face. 
Can you share your twitter/hugging face handles so we give you credit for this idea?

**Comment by mobicham** [comment]
I just added support for custom quantization config: https://github.com/mobiusml/hqq/commit/37502bea31f2969c6680c0c4a88ca74b3bb234a5

I also tried it on Llama-13B with mixed 2-bit/4-bit by using 4-bit for the `self_attn.v_proj` layers and 2-bit for the rest, not a huge improvement but a bit better for an additional 0.2 GB:
| Method      | PPL | Memory     |
| :---        |    :----:   |          ---: |
HQQ_2bit_g16_z_s128 (all layers)	| 5.7 GB	| 6.371	
HQQ_4-bit_g64_z_s256 (v_proj)/ HQQ_2bit_g16_z_s128 (rest) |	5.9 GB |	6.121	

We are gonna publish the new Mixtral models later today: [Base](https://huggingface.co/mobiuslabsgmbh/Mixtral-8x7B-v0.1-hf-attn-4bit-moe-2bit-HQQ), [Instruct](https://huggingface.co/mobiuslabsgmbh/Mixtral-8x7B-Instruct-v0.1-hf-attn-4bit-moe-2bit-HQQ).

Cheers!
